darts/models/forecasting/query_selector.py

```diff
@@ -0,0 +1,650 @@
+            self.feed_forward_network_enc = FeedForward
+            self.feed_forward_network_dec = FeedForward
+        else:
+            raise_if_not(
+                self.feed_forward in GLU_FFN,
+                f"'{self.feed_forward}' is not in {GLU_FFN + ['default']}",
+            )
+            # use glu variant feedforward layers
+            # 4 is a commonly used feedforward multiplier
+            self.feed_forward_network_enc = getattr(glu_variants, self.feed_forward)
+            self.feed_forward_network_dec = getattr(glu_variants, self.feed_forward)
+
+        self.feed_forward_network_enc = self.feed_forward_network_enc(
+            dim_val, dim_val * 4, dropout=0.1
+        )
+        self.feed_forward_network_dec = self.feed_forward_network_dec(
+            dim_val, dim_val * 4, dropout=0.1
+        )
+
+        # Initiate encoder and Decoder layers
+        self.encs = []
+        for i in range(n_encoder_layers):
+            self.encs.append(
+                EncoderLayer(
+                    dim_val,
+                    dim_attn,
+                    feed_forward=self.feed_forward_network_enc,
+                    n_heads=n_heads,
+                    attn_type=enc_attn_type,
+                    norm=self.layer_norm,
+                )
+            )
+        self.encs = InferenceModuleList(self.encs)
+        self.decs = []
+        for i in range(n_decoder_layers):
+            self.decs.append(
+                DecoderLayer(
+                    dim_val,
+                    dim_attn,
+                    feed_forward=self.feed_forward_network_dec,
+                    n_heads=n_heads,
+                    attn_type=dec_attn_type,
+                    norm=self.layer_norm,
+                )
+            )
+        self.decs = InferenceModuleList(self.decs)
+        self.pos = PositionalEncoding(dim_val)
+
+        self.enc_dropout = MonteCarloDropout(p=dropout)
+
+        # Dense layers for managing network inputs and outputs
+        self.enc_input_fc = nn.Linear(self.input_dim, dim_val)
+        self.dec_input_fc = nn.Linear(self.input_dim, dim_val)
+        self.out_fc = nn.Linear(
+            dec_seq_len * dim_val,
+            self.output_chunk_length * self.output_dim * self.nr_params,
+        )
+
+        self.debug = debug
+
+    def forward(self, x_in: tuple):
+        x, _ = x_in
+        # x: [Batch, Input length, Channel]
+
+        # encoder
+        e = self.encs[0](self.pos(self.enc_dropout(self.enc_input_fc(x))))
+
+        for enc in self.encs[1:]:
+            e = enc(e)
+        if self.debug:
+            print(f"Encoder output size: {e.shape}")
+        # decoder
+        decoded = self.dec_input_fc(x[:, -self.dec_seq_len :])
+
+        d = self.decs[0](decoded, e)
+        for dec in self.decs[1:]:
+            d = dec(d, e)
+
+        # output
+        x = self.out_fc(d.flatten(start_dim=1))
+
+        x = torch.reshape(x, (x.shape[0], -1, self.output_dim))
+        x = x.view(
```

Comment:
I see they also do it like that in the reference implementation, and we also do it like that in our current `Transformer` implementation, but I'm not sure that's the best way.
Currently, in a single forward pass we output one large vector of size `self.output_chunk_length * self.output_dim * self.nr_params`, which we reshape in order to get a time series of length `self.output_chunk_length`. However, if you look at the transformer paper, I believe a more natural way to do this would be to auto-regressively output the next time steps; i.e., instead of outputing `output_chunk_length` all at once, output first `t+1` in a forward pass, then feed this as an input to the decoder and output `t+2`, etc. So perform `output_chunk_length` forward passes. I believe this might be better as it would avoid depending solely on the last FC layer to capture everything needed about an arbitrary number of future time steps. That might come at the price of increase run time for inference though. Perhaps we could leave this as an option?